### PR TITLE
parse embedded json in log/MESSAGE fields

### PR DESCRIPTION
**Pull request by richm**
parse embedded json in log/MESSAGE fields

### CODE REVIEW

**Comment by richm** [comment]
flake?
/test logging

File: hack/testing/rsyslog/parse_json.rulebase [hunk 1]
```diff
@@ -1 +1,2 @@
-rule=:%payload:json%
+version=2
+rule=:%.:json%
```
Comment (nhosoi) [inline hack/testing/rsyslog/parse_json.rulebase]: Wow.  This is very simple and clean.
I have one question.  If a log like this is given:
`{"log":"{\"message\":\"some_message\"}", "others":"{\"foo\":\"bar\"}", "top":"level"}`
the parsed result would be
`{"message":"some_message", "foo":"bar", "top":"level"}`
?  Or are there any way to specify just process the value of "log"?  (I guess there's no such use case for now, though... :)

Comment (richm) [reply hack/testing/rsyslog/parse_json.rulebase]: This makes it parse only the value of `$!log` - it was already in there so it didn't have to change, only the rulebase file:
```
          action(type="mmnormalize" ruleBase="/etc/rsyslog.d/viaq/parse_json.rulebase" variable="$!log") 
```
so the parsed result is
```
{"message":"some_message", "others":"{\"foo\":\"bar\"}", "top":"level","log":"{\"message\":\"some_message\"}"}
```
I still need to add in the code to delete `$!log` or `$!MESSAGE` if it was parsed correctly.

Comment (nhosoi) [reply hack/testing/rsyslog/parse_json.rulebase]: or keep it with naming the key to "original_log":"{\"message\":\"some_message\"}?

Comment (richm) [reply hack/testing/rsyslog/parse_json.rulebase]: or yes, keep it
```
set $!pipeline_metadata.collector.original_raw_message = $!log;
unset $!log;
```
or something like that

Comment (nhosoi) [reply hack/testing/rsyslog/parse_json.rulebase]: The change is going to be in the CI test rsyslog config?  If so, should I wait for another commit to give lgtm?

Comment (richm) [reply hack/testing/rsyslog/parse_json.rulebase]: yes, wait

**Comment by nhosoi** [comment]
Hi Rich, for some reason, I cannot make the rulebase work with variable...  This is parse_json.rb copied from this pr.
```
version=2
rule=:%.:json%
```
If I set variable="$!log" as in the CI test
`action(type="mmnormalize" ruleBase="/etc/rsyslog.d/parse_json.rb" variable="$!log")`
the parsed output says original message is empty.
`{ "metadata": { "filename": "\/var\/log\/containers\/loader-0.log", "fileoffset": "12124" }, "originalmsg": "", "unparsed-data": "" }`
If I don't set variable, then the output logs all.
`{ "metadata": { "filename": "\/var\/log\/containers\/loader-0.log", "fileoffset": "13217" }, "log": "{\"message\":\"loader seq - 24e3d9edf32142d6abf46c9d653cbc2e - 0000000003 - llx0hp6tnzfbv...3lil1e81dy\"}\n", "stream": "stdout", "time": "2018-07-24T21:42:06.968446829Z" }`
I debugged in gdb, but for some reason when getting the value of "log", the parent json says there's no such child...  I experimentally did mmjsonparse before mmnormalize, but it did not help either. :(
```
action(type="mmjsonparse" cookie="")
action(type="mmnormalize" ruleBase="/etc/rsyslog.d/parse_json.rb" variable="$!log")
```
I must be doing something stupid, but I cannot find it...

**Comment by richm** [comment]
what version of rsyslog are you using?

**Comment by richm** [comment]
hold - looks like a problem with journal - tests with log driver json-file work fine, broken with log driver journald

**Comment by richm** [comment]
@nhosoi did you figure it out?  try this
```
action(type="mmjsonparse" cookie="")
action(type="omfile" file="/var/log/rsyslog-debug-1.log" template="RSYSLOG_DebugFormat")
action(type="mmnormalize" ruleBase="/etc/rsyslog.d/parse_json.rb" variable="$!log")
action(type="omfile" file="/var/log/rsyslog-debug-2.log" template="RSYSLOG_DebugFormat")
```

**Comment by nhosoi** [comment]
Well, I haven't...  I was working on the skipempty side... :p

The 2 rsyslog-debug logs are almost identical following the Debug line [1].  (Note: both $. and $/ are empty.)
```
# diff rsyslog-debug-[12].log
8c8
< $!:{ "metadata": { "filename": "\/var\/log\/containers\/loader-0.log", "fileoffset": "14611" } }
---
> $!:{ "metadata": { "filename": "\/var\/log\/containers\/loader-0.log", "fileoffset": "14611" }, "originalmsg": "", "unparsed-data": "" }
```
[1]
```
Debug line with all properties:
FROMHOST: '', fromhost-ip: '', HOSTNAME: 'momo7', PRI: 133,
syslogtag 'docker', programname: 'docker', APP-NAME: 'docker', PROCID: '-', MSGID: '-',
TIMESTAMP: 'Jul 26 14:16:34', STRUCTURED-DATA: '-',
msg: '{"log":"{\"message\":\"Test message 0-0\",\"channel\":\"webservice\",\"test\":{},\"log_level\":\"INFO\"}", "key0":[], "key1":{}, "key2":"", "a0":[{"k0":"v0"},{},{"k1":"v1"},[],{"k2":"v2"}],"log_level":"DEBUG", "stream": "stdin","time": "2019-05-03T17:43:26.87+00:20", "nested":{"n0":"nval0","n1":""}}'
escaped msg: '{"log":"{\"message\":\"Test message 0-0\",\"channel\":\"webservice\",\"test\":{},\"log_level\":\"INFO\"}", "key0":[], "key1":{}, "key2":"", "a0":[{"k0":"v0"},{},{"k1":"v1"},[],{"k2":"v2"}],"log_level":"DEBUG", "stream": "stdin","time": "2019-05-03T17:43:26.87+00:20", "nested":{"n0":"nval0","n1":""}}'
inputname: imfile rawmsg: '{"log":"{\"message\":\"Test message 0-0\",\"channel\":\"webservice\",\"test\":{},\"log_level\":\"INFO\"}", "key0":[], "key1":{}, "key2":"", "a0":[{"k0":"v0"},{},{"k1":"v1"},[],{"k2":"v2"}],"log_level":"DEBUG", "stream": "stdin","time": "2019-05-03T17:43:26.87+00:20", "nested":{"n0":"nval0","n1":""}}'
```

**Comment by richm** [comment]
So the problem is that the first mmjsonparse isn't doing anything . . .

**Pull request merged**